x/staking/keeper/val_state_change.go

```diff
@@ -164,12 +164,12 @@ func (k Keeper) ApplyAndReturnValidatorSetUpdates(ctx sdk.Context) (updates []ab
 
 		copy(valAddrBytes[:], valAddr[:])
 		oldPowerBytes, found := last[valAddrBytes]
-		newPower := validator.ConsensusPower()
+		newPower := validator.ConsensusPower(k.PowerReduction(ctx))
 		newPowerBytes := k.cdc.MustMarshalBinaryBare(&gogotypes.Int64Value{Value: newPower})
 
 		// update the validator set if power has changed
 		if !found || !bytes.Equal(oldPowerBytes, newPowerBytes) {
-			updates = append(updates, validator.ABCIValidatorUpdate())
+			updates = append(updates, validator.ABCIValidatorUpdate(k.PowerReduction(ctx)))
```

Comment:
Reading power reduction from keeper several times from KVStore is not recommended as it will increase gas usage. Recommend to use a local variable to reduce gas.

Reply:
Agreed, I think locally caching this value with a nice name will also help with readability

Reply:
Technically, it's not costing anyone any gas, because its in the endblocker, but other than that, also agree.  Might as well cache outside the for loop to save unnecessary reads.